### ISSUE TITLE
React LoginCallback Component: "AuthSdkError: Could not load PKCE codeVerifier from storage. This may indicate the auth flow has already completed or multiple auth flows are executing concurrently.

### ISSUE THREAD

**Issue by AndreyZSneg** [okta/okta-oidc-js#898]
## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [ ] okta-angular
- [ ] oidc-middleware
- [x] okta-react
- [ ] okta-react-native

## I'm submitting a:

- [x] Bug report  
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior

Currently our react application is set up to immediately redirect the user to a protected route (the main app page), which redirects them to the Okta login (when not already authenticated). After being redirected back, it should load the main page.

In Chrome and Firefox everything works as expected, but in Edge (v41.16299, note that in Edge v80+ everything works as expected) we are getting:

`AuthSdkError: Could not load PKCE codeVerifier from storage. This may indicate the auth flow has already completed or multiple auth flows are executing concurrently.`

In the HTML, while still sitting at the /implicit/callback route. Clearing cache and re-login does not help to solve this problem.

## Expected behavior

The authentication flow should work without any errors.

## Minimal reproduction of the problem with instructions

In MS Edge v41

1. Launch application
2. Log in through Okta
3. Successful login redirects to /implicit/callback page (with the LoginCallback), returns the AuthSdkError.

## Environment

- Package Version: 3.0.6
- Browser: Edge 41.16299
- OS: Windows 10
- Node version (`node -v`): 13.5.0
- Other: React 16.9.0

**Comment by aarongranick-okta** [comment]
@AndreyZSneg The most common cause of this error is multiple callback components loaded on a single request or multiple loads of the callback route. We recommend wrapping your routes in the `<Switch>` component and setting the `exact` keyword on your callback route. It may be possible that the callback is being loaded twice because no URI is being set before the redirect. Our `<SecureRoute>` component saves the current URI before beginning the login flow, which will proceed:

```
current -> okta login -> callback -> current
```

Something may have gone wrong and instead of setting location from callback -> current it went from callback -> callback.

My suggestion is to set a breakpoint in the callback component (or add a `debugger` statement) and load the page with the debugger running. If the callback component loads twice, the second load will probably be an effect of the source of the error.

**Comment by AndreyZSneg** [comment]
Hello @aarongranick-okta. Thanks for your reply. Proposed fix with `<Switch>`, `exact` and debugging did not solve our issue. Some code samples:
`
const authService = new AuthService({
  issuer: 'ourIssuerLink',
  clientId: 'ourClientId',
  redirectUri: window.location.origin + '/implicit/callback',
  scope: ['email', 'profile', 'openid'],
  pkce: true,
});

const TestApp = () => (
  <Security authService={authService}>
    <div>Okta test App</div>
    <BrowserRouter>
      <Switch>
        <SecureRoute
          path={homeUrl}
          exact={true}
          component={() => <>test</>}
        />
        <Route exact={true} path="/implicit/callback" component={LoginCallback} />
      </Switch>
    </BrowserRouter>
  </Security>
);

render(<TestApp />, root);
`
debugging did not identify any multiple component calls.

Closer debugging of the library showed that:
Okta Auth JavaScript SDK uses session and local storage to temporally save PKCE code verifier.

These storages are partitioned by security zones in the IE11 and older versions of Microsoft Edge browser. When the application is redirected across zones, the session storage and local storage are cleared. Specifically, the session storage is cleared in the regular browser navigation, and both the session and local storage are cleared in the InPrivate mode of the browser.

Therefore, the mentioned error can be received opening application in IE11 trying to get PKCE code verifier after it was cleared from storage upon navigation between application and Okta.

Could you please advice on solution.

**Comment by AndreyZSneg** [comment]
Hi guys, any updates on this issue?

**Comment by swiftone** [comment]
@AndreyZSneg - Thanks for the detailed debugging.  I'll check in with the team to see if we have any update.

**Comment by swiftone** [comment]
@AndreyZSneg - Looks like there's a lot of changes happening with PKCE claims and the like, but nothing that yet solves the issue with older browsers you've outlined.  

One workaround you can try is to bypass the redirect that is giving you the PKCE error.  You can see where this is done in some sample code here: https://github.com/okta/samples-js-react/pull/137

**Comment by arthyn** [comment]
Hey @swiftone can you maybe expand upon what the implications are of doing the bypass?

**Comment by aarongranick-okta** [comment]
@arthyn We have two samples for `okta-react` available here: https://github.com/okta/samples-js-react

One sample, `okta-hosted` shows how to redirect to Okta for the signin flow. The other sample, `custom-login` shows how to provide the signin flow within your own application using the Okta signin widget: https://github.com/okta/okta-signin-widget

The widget can be added to your application either by including the NPM module or by adding a script tag on your page.  Either way, embedding the widget within your app will allow your users to complete the signin flow without redirecting to Okta. This should avoid the problems you are seeing due to the redirect.

Many customers prefer using the embedded widget because it is a more seamless experience and you can customize it with your own CSS. Other HTML on the page can provide additional instructions.  The main implication is that you will be using a specific version of the widget. As time goes on you will want to manually upgrade the version to keep up with the latest improvements and features. Redirecting to the Okta hosted widget will always load the most recent version.

**Comment by MikeSpencer42** [comment]
We are using the okta-vue implementation and are seeing the same issue in MS Edge v44. The issue appears to be intermittent however. I tried porting the code from the pull request posted by @swiftone but that resulted in an error stating "Unable to get user profile. Please log out and try again." Are there also setting changes with the Okta admin console that go along with these code changes?

**Comment by andrewpaulinoev** [comment]
Any updates to this? it seems like using the hosted version of Okta with React produces this error a lot. I'd rather not have to recreate the widget in my application to use the SSO Functionality.

**Comment by aarongranick-okta** [comment]
@andrewpaulinoev We have made several updates to address this issue. If you have the latest version installed (plus dependencies, it may be necessary to remove your node_modules and any ".lock" files and reinstall), the issue should corrected (or compensated for) in most cases. The most common cause of this error is some form of concurrency, basically a race condition. With this version of the SDK, token renew occurs when the token is read (`tokenManager.get("accessToken")`). If token renew is occurring at the same time that a normal auth flow is being triggered, the two flows may compete with each other and overwrite each other's storage. It is important to avoid any type of actions with the token (including read) when processing the CALLBACK. Make sure you don't have any code which is trying to determine authorization status or retrieve the user's profile while processing the login callback from the Okta redirect.

Another option is to avoid redirect entirely by embedding the Okta signin widget in your application. This example shows how this can be done: https://github.com/okta/samples-js-react/tree/master/custom-login

**Comment by d12santosh** [comment]
was able to resolve this issue by moving logic of storing information in session storage to the level post completion of security route configuration, hope it helps

**Comment by clearwaterstream** [comment]
@d12santosh can you please provide a rough sample or pseudo code of the solution? I'm having trouble visualizing it. It would be much appreciated.

P.S. I was able to work around it by not using the `core-js/features/string/includes` polyfill. I had to also remove `okta-pkce-storage` values from session storage prior to login.

See sample here:

https://github.com/clearwaterstream/okta-ie11-issue/tree/working

**Comment by kellengreen** [comment]
Since upgrading to `@okta/okta-react:6.0.0`, I'm now running into this regularly.

**Comment by clearwaterstream** [comment]
We've opened a ticket with Okta and the workaround that was provided by Okta support engineer was this:

```javascript
const oktaConfig = {
    token: {
        storageTypes: [
            'localStorage',
            'sessionStorage',
            'cookie'
        ],
        useMultipleCookies: true // puts each token in its own cookie
    },
    cache: {
        storageTypes: [
            'localStorage',
            'sessionStorage',
            'cookie'
        ]
    },
    transaction: {
        storageTypes: [
            'cookie'
        ]
    }
};
```

https://github.com/okta/okta-auth-js

We tested in IE 11 21H1 and earlier versions and it seem to work. We only enabled it only on IE browsers only. For other browsers we did not fiddle with the settings.

**Comment by bengwall** [comment]
I'm having this same issue also with v6.0.0.

**Comment by aarongranick-okta** [comment]
@kellengreen @bengwall 

Is the issue seen on IE 11 browser only?

**Comment by kellengreen** [comment]
All browser's for me.

**Comment by dean-dot-git** [comment]
@aarongranick-okta , any word on the fix for this?  All browsers for me, I'm using okta-auth-js v5.1.1 and okta-react v6.0.0.

Looking to get this resolved quickly.

**Comment by aarongranick-okta** [comment]
@kellengreen @bengwall 
- Are you using an embedded (self-hosted) signin widget, or does your app redirect to Okta-hosted login page? 
- Can you share your config (without private information), especially related to storage or widget?

**Comment by kellengreen** [comment]
We're using `app redirect to Okta-hosted login page`.

with a very basic `config`

```js
export default new OktaAuth({
  issuer: OKTA_ISSUER,
  clientId: OKTA_CLIENT,
  redirectUri: `${globalThis.location.origin}/implicit/callback`,
})

**Comment by aarongranick-okta** [comment]
@kellengreen That looks fine. Maybe the issue is with how the callback is being handled. Are you using the LoginCallback component included with the okta-react SDK? Are you putting in a route with a <Switch> ? We have seen problems with react router where the login callback (or another component meant for unauthenticated users) renders on the callback. The expectation is that the LoginCallback is the ONLY component to render on the route, and should render only once. When the callback logic is complete it will redirect to the "original" route, user should not be left on `/login/callback`. rendering this route after callback is handled could also cause this error about PKCE storage.

**Comment by kellengreen** [comment]
Yes we're using the `LoginCallback` component, and yes it's wrapped around `Switch`.

I should reiterate our setup worked just fine until the upgrade.

**Comment by finn-orsini** [comment]
Hi folks - I am also seeing this issue intermittently. Our app is using `@okta/okta-react@5.1.2` and `@okta/okta-auth-js@4.9.1`. We are not using the sign in widget - it is a full redirect.  The routes are as recommended, where `LoginCallback` is imported directly from `@okta/okta-react`, i.e. 

```js

  <Router basename="<baseName>">
      <Security oktaAuth={oktaAuth} restoreOriginalUri={restoreOriginalUri}>
        <Switch>
          <Route
            exact
            path="/login/callback"
            render={(props) => (
              <LoginCallback {...props} errorComponent={Error} />
            )}
          />
          <Route/>
        </Switch>
      </Security>
    </Router>
```

and our okta config is very simple: 


```js
const oktaAuth = new OktaAuth({
  issuer,
  clientId,
  redirectUri,
  pkce: true,
});

```


The behavior is not consistently reproducible, nor specific to one browser - but most of my users use Chrome.

Is there anything in particular you would recommend logging to help track down what is causing this issue when it happens?

**Comment by dean-dot-git** [comment]
Following up with my information/config:

- Hosted login.  Intentionally done so we don't have to constantly update the SIW.
- Okta's LoginCallback (not a custom one).

Versions:
@okta/okta-react:  6.0.0
@okta/okta-auth-js: 5.2.0

Config:
![image](https://user-images.githubusercontent.com/86485757/125214737-35205480-e27e-11eb-8506-450e145ff700.png)

Switch (there is a "BrowserRouter" upstream one level):
![image](https://user-images.githubusercontent.com/86485757/125214805-a5c77100-e27e-11eb-9e13-5f2d64a8b94b.png)

Let me know if I can help in any way to get this solved.  The application I'm building is in development, so I have latitude to do what I need to make changes pretty easily.

**Comment by chefong** [comment]
> Since upgrading to `@okta/okta-react:6.0.0`, I'm now running into this regularly.

@kellengreen were you facing this issue before 6.0.0 and did you have a workaround for it back then? Just wondering because i've also encountered it on a previous version (forgot which one though)

**Comment by kellengreen** [comment]
No, this started for us immediately after upgrading to `6.0`.

**Comment by chefong** [comment]
@kellengreen Which versions of `okta-react` and `okta-auth-js` did you have working prior to the upgrade, if i may ask?

**Comment by kellengreen** [comment]
When I'm back on VPN, I will check for you tomorrow.

@aarongranick-okta whats the status of this issue? My initial reaction is to simply downgrade, but it sounds like this made it's way into 5.x as well.

**Comment by kellengreen** [comment]
This was our upgrade path:

**okta-auth-js** - `4.9.2` > `5.1.1`
**okta-react** - `5.1.2` > `6.0.0`

**Comment by mpelham** [comment]
@kellengreen did you wind up downgrading?

**Comment by kellengreen** [comment]
I ended up forcing `signOut()` on any login errors for now.

Hopefully this will just be temporary, as this leaves us with no way of displaying "real" errors to end users.

```js
() => {
  const { oktaAuth } = useOktaAuth()
  return (
    <LoginCallback
      errorComponent={() => {
        oktaAuth.signOut("/")
      }}
    />
  )
}
```

**Comment by kellengreen** [comment]
I've documented the steps to reproduce [here](https://github.com/kellengreen/okta).

**Comment by shuowu** [comment]
@kellengreen Thanks for providing the detailed repro steps! If the error is caused by using `code` query param in your url. There is a relevant thread here: https://github.com/okta/okta-auth-js/issues/474

tl:dr: the protection logic was introduced to protect the PKCE flow against race condition. We'll revisit the logic to see if any improvement can be done.

Detailed explanation: https://github.com/okta/okta-auth-js/issues/474#issuecomment-696268565

**Comment by kellengreen** [comment]
I understand, `code=` was just an easier way to reproduce locally.

In the real world it goes something like this:

- User comes back to the site after their tokens have expired.
- `<SecureRoute>` triggers the login process.
- User is sent back to `<LoginCallback>` which hard fails with `Could not load PKCE codeVerifier from storage`.

**Comment by kellengreen** [comment]
Due to another probably related [bug](https://github.com/okta/okta-react/issues/148) I can no longer recommend my solution from above. I've needed to add a fallback for when `signOut` stalls out.

```js
const d = () => {
  const { oktaAuth } = useOktaAuth();
  return (
    <LoginCallback
      errorComponent={() => {
        const id = setTimeout(() => {
          localStorage.clear()
          window.location = '/'
        }, 2000);
        oktaAuth.signOut().finally(() => {
          clearTimeout(id);
        });
        return null;
      }}
    />
  );
};
```

OKTA folks do we need to move this to https://github.com/okta/okta-react/issues?

**Comment by kellengreen** [comment]
Tested using `@okta/okta-auth-js:5.2.3`. The results are the same.

**Comment by denysoblohin-okta** [comment]
> OKTA folks do we need to move this to https://github.com/okta/okta-react/issues?

No need. We are moving issues from `okta-oidc-js` to `okta-react`, but this issue has useful conversation for investigation.
We are aware of this issue. Internal ref: OKTA-414847

**Comment by chefong** [comment]
@denysoblohin-okta @shuowu  any plans on prioritizing this issue in the near future?

**Comment by Zuhaib-SB** [comment]
> @denysoblohin-okta @shuowu any plans on prioritizing this issue in the near future?

I used the latest okta react package (6.1.0) and I no longer get this issue. I think it's been fixed (for me at least)

**Comment by kellengreen** [comment]
Using `@okta/okta-auth-js=5.4.2` and `@okta/okta-react": "6.1.0` yield the same results for me. I've updated [my repo](https://github.com/kellengreen/okta) to use the latest releases if anyone else would like to test.

**Comment by Zuhaib-SB** [comment]
Yes still seeing this bug from time to time actually.

**Comment by dean-dot-git** [comment]
I'm continuing to see this as @kellengreen still does using the same versions he did.  I'm working with react 6.2 and auth-js 5.5 now.  @kellengreen are you still seeing it with these new versions?

**Comment by kellengreen** [comment]
Yep still broke on `5.5.0/6.2.0`. I've updated [my repo](https://github.com/kellengreen/okta) as well.

**Comment by oleksandrpravosudko-okta** [comment]
Investigating code paths that can lead to `Could not load PKCE codeVerifier from storage` error did not reveal explicit code issues in the SDKs.
I am going to close this as non-reproducible. In case this error is still observable in latest SDKs versions, please log a new issue with the following details provided:
* type of storage being used for transaction data
* scopes value
* whether app has multiple tabs open

As a side note, we plan to enhance logging in the SDKs to make troubleshooting such issues a bit easier.
